api/odl.py

```diff
@@ -786,6 +765,82 @@ class ODLImporter(OPDSImporter):
     # about the license.
     LICENSE_INFO_DOCUMENT_MEDIA_TYPE = 'application/vnd.odl.info+json'
 
+    @classmethod
+    def parse_license(
+            cls,
+            identifier: str,
+            total_checkouts: Optional[int],
+            concurrent_checkouts: Optional[int],
+            expires: Optional[datetime.datetime],
+            checkout_link: Optional[str],
+            odl_status_link: Optional[str],
+            do_get: Callable
+    ) -> Optional[LicenseData]:
+        remaining_checkouts = None
+        available_concurrent_checkouts = None
+
+        while True:
+            if total_checkouts is not None:
+                total_checkouts = int(total_checkouts)
+
+                if total_checkouts <= 0:
+                    logging.info(
+                        f"License # {identifier} expired since "
+                        f"the total number of checkouts is {total_checkouts}"
+                    )
+                    break
+
+            if expires:
+                if not isinstance(expires, datetime.datetime):
+                    expires = dateutil.parser.parse(expires)
+
+                expires = util.datetime_helpers.to_utc(expires)
+                now = util.datetime_helpers.utc_now()
+
+                if expires <= now:
+                    logging.info(
+                        f"License # {identifier} expired at {expires} (now is {now})"
+                    )
+                    break
+
+            if odl_status_link:
+                status_code, _, response = do_get(
+                    odl_status_link, headers={}
+                )
+
+                if status_code < 400:
```

Comment:
You're absolutely right, I added an `else` statement logging the error and breaking out of the cycle. I'm not sure about the `200` though. Do you think it's possible that a status link returns a 3xx code?

Reply:
I thought that `requests` library was more intelligent. Would you like to change it to the following?
```
if status_code < 300:
```

[The documentation](https://github.com/readium/readium-lcp-server/wiki/LCP-Status-Server-API#return-a-license-status-document) says that the lsdserver can return 200+ codes.

Reply:
Even so, I suspect that we'd get 200 or 201 in the only two cases that return the data we really handle. I suppose we could deal with a 203, since we would expect to get the full document back, but I don't think we would expect a non-authoritative response when we're trying to find out the license status.